packages/memoize/src/index.ts

```diff
@@ -1,8 +1,12 @@
-export default function memoize<V>(fn: (arg: string) => V): (arg: string) => V {
-  const cache: Record<string, V> = Object.create(null)
+function memoize<V>(fn: (arg: string) => V): (arg: string) => V
+function memoize<K extends PropertyKey, V>(fn: (arg: K) => V): (arg: K) => V
```

Comment:
I wonder if it isn't better to reorder those as overloads are always matched from the top to the bottom. OTOH when using some conditional types (similar to `Parameters`, `ReturnType` etc) only the last overload gets picked 🤔 
```suggestion
function memoize<K extends PropertyKey, V>(fn: (arg: K) => V): (arg: K) => V
function memoize<V>(fn: (arg: string) => V): (arg: string) => V
```

Perhaps... even though this is somewhat better, maybe we shouldn't adjust this now? Or we should just adjust it and call it a day? I'm kinda on the fence here. 

Could you describe why this was needed? I know it's probably related to `isPropValid` or `shouldForwardProp` but I didn't have time to analyze this.

Reply:
I think your code change makes sense, so I committed it.

As for why this was needed, I was getting a dtslint error from `typescript@4.7` on [this line](https://github.com/emotion-js/emotion/blob/9ca22c6c23e9effa086d161a9b0ae1c645686680/packages/styled/types/tests-base.tsx#L114). If you look at [the definition of `shouldForwardProp`](https://github.com/emotion-js/emotion/blob/9ca22c6c23e9effa086d161a9b0ae1c645686680/packages/styled/types/base.d.ts#L28), the `prop` argument has type `PropertyKey` a.k.a. `string | number | symbol`. However, `isPropValid` was only accepting `string` inputs, hence my changes to `isPropValid` and `memoize`.

Reply:
I wonder if we shouldn't change this the other way around. It's just so unlikely for people to have number/symbol props... 🤔

Reply:
Yeah, I agree that it is extremely unlikely.... 

I also just realized that using `PropertyKey` as the argument of `shouldForwardProp` could be annoying for consumers of Emotion since they will have to do a `typeof prop === 'string'` check before being able to use any string-specific methods.

I am in favor of reverting the `memoize` and `isPropValid` changes and just making everything `string`. If you approve I will go ahead and do it.

Reply:
Perhaps a constraint to `Props` could be introduced - so TS would understand that only `string` is a valid property key there. That would probably be best but I'm not sure what impact this would have on the rest of our definitions. Alternatively, we can use smth like `keyof Props & string` to scope `isPropValid` to work on string-only keys.

I will take a look at this soon to check out what is better here

Reply:
Adding a contraint to `Prop` seems rather hard to pull off. The second suggestion can be applied using this patch:
```diff
diff --git a/packages/styled/types/base.d.ts b/packages/styled/types/base.d.ts
index c39200c7..4383b0ca 100644
--- a/packages/styled/types/base.d.ts
+++ b/packages/styled/types/base.d.ts
@@ -16,16 +16,16 @@ export { ComponentSelector, Interpolation }
 /** Same as StyledOptions but shouldForwardProp must be a type guard */
 export interface FilteringStyledOptions<
   Props,
-  ForwardedProps extends keyof Props = keyof Props
+  ForwardedProps extends keyof Props & string = keyof Props & string
 > {
   label?: string
-  shouldForwardProp?(propName: PropertyKey): propName is ForwardedProps
+  shouldForwardProp?(propName: string): propName is ForwardedProps
   target?: string
 }
 
 export interface StyledOptions<Props> {
   label?: string
-  shouldForwardProp?(propName: PropertyKey): boolean
+  shouldForwardProp?(propName: string): boolean
   target?: string
 }
 
@@ -118,7 +118,8 @@ export interface CreateStyledComponent<
 export interface CreateStyled {
   <
     C extends React.ComponentClass<React.ComponentProps<C>>,
-    ForwardedProps extends keyof React.ComponentProps<C> = keyof React.ComponentProps<C>
+    ForwardedProps extends keyof React.ComponentProps<C> &
+      string = keyof React.ComponentProps<C> & string
   >(
     component: C,
     options: FilteringStyledOptions<React.ComponentProps<C>, ForwardedProps>
@@ -147,7 +148,8 @@ export interface CreateStyled {
 
   <
     C extends React.ComponentType<React.ComponentProps<C>>,
-    ForwardedProps extends keyof React.ComponentProps<C> = keyof React.ComponentProps<C>
+    ForwardedProps extends keyof React.ComponentProps<C> &
+      string = keyof React.ComponentProps<C> & string
   >(
     component: C,
     options: FilteringStyledOptions<React.ComponentProps<C>, ForwardedProps>
@@ -168,7 +170,8 @@ export interface CreateStyled {
 
   <
     Tag extends keyof JSX.IntrinsicElements,
-    ForwardedProps extends keyof JSX.IntrinsicElements[Tag] = keyof JSX.IntrinsicElements[Tag]
+    ForwardedProps extends keyof JSX.IntrinsicElements[Tag] &
+      string = keyof JSX.IntrinsicElements[Tag] & string
   >(
     tag: Tag,
     options: FilteringStyledOptions<JSX.IntrinsicElements[Tag], ForwardedProps>

```

It's not great, but it is workable and I wonder if we shouldn't "go for it".  🤔

Reply:
LGTM, I have made this change in a485db3ca57c03642986f879ddcf3814c9c1fc5e which I am about to push. Made the same change to the `@emotion/native` types. Fixing something else in this branch and then I will push.